### ISSUE TITLE
BinaryInputDetector Discussion

### ISSUE THREAD

**Issue by authwork** [Trusted-AI/adversarial-robustness-toolbox#766]
Hello!
Would you please show me some difference about BinaryInputDetector and a normal classifier?
In the example "detection_adversarial_samples_cifar10".
I compare three pieces of codes:
```
path = get_file('BID_eps=0.05.h5',extract=False, path=ART_DATA_PATH,
                url='https://www.dropbox.com/s/cbyfk65497wwbtn/BID_eps%3D0.05.h5?dl=1')
detector_model = load_model(path)
detector_classifier = KerasClassifier(clip_values=(-0.5, 0.5), model=detector_model, use_logits=False)
detector = BinaryInputDetector(detector_classifier)
detector.fit(x_train_detector, y_train_detector, nb_epochs=20, batch_size=20)
```
```
path = get_file('BID_eps=0.05.h5',extract=False, path=ART_DATA_PATH,
                url='https://www.dropbox.com/s/cbyfk65497wwbtn/BID_eps%3D0.05.h5?dl=1')
detector = load_model(path)
detector.compile(optimizer='rmsprop',loss='binary_crossentropy',metrics=['accuracy'])
detector.fit(x_train_detector, y_train_detector, epochs=20, batch_size=20)
```
```
detector = keras.applications.DenseNet201(weights=None, input_shape=(32, 32, 3), classes=2)
detector.compile(optimizer='rmsprop',loss='binary_crossentropy',metrics=['accuracy'])
detector.fit(x_train_detector, y_train_detector, epochs=20, batch_size=20)
```
The first two have very similar accuracy, said 0.7~0.8. The third has only 0.5. But the prediction varies a lot:

First solution:
```
Adv ->
Flagged: 37
Not:63
Original ->
Flagged: 0
Not:100
```
Second solution:
```
Adv ->
Flagged: 50
Not:50
Original ->
Flagged: 0
Not:50
```
Third solution:
```
Adv ->
Flagged: 100
Not:0
Original ->
Flagged: 100
Not:0
```
Currently, I guess it may be caused by last few layers:
```
_________________________________________________________________________________________________
average_pooling2d_1 (AveragePoo (None, 1, 1, 64)     0           dropout_1[0][0]                  
__________________________________________________________________________________________________
flatten_1 (Flatten)             (None, 64)           0           average_pooling2d_1[0][0]        
__________________________________________________________________________________________________
classifier (Dense)              (None, 2)            130         flatten_1[0][0]                  
==================================================================================================
```

update：

When I use the following code, it works not so bad:
```
# I just replace the output of the classifier_model
layers = [layer for layer in classifier_model.layers]
num_layers = len(layers)
output = keras.layers.Dense(2)(layers[num_layers-2].output)
detector = keras.Model(inputs=layers[0].input, outputs=output)
detector.summary()
```
acc = 0.79
```
Adversarial test data (first 100 images):
Flagged: 81
Not flagged: 19
Original test data (first 100 images):
Flagged: 54
Not flagged: 46
```

**Comment by beat-buesser** [comment]
Hi @authwork Thank you very much for raising and investigating this issue!

It seem there is no difference between `BinaryInputDetector` and a normal classifier. We might have to revisit `BinaryInputDetector` in the near future to update it because it has been implemented before the introduction of `art.estimators`.

We also need to update the notebook accordingly and add a warning that detectors can be fooled too.
